### ISSUE TITLE
Rule proposal: no-unused-assign

### ISSUE THREAD

**Issue by ashmind** [eslint/eslint#10885]
**Please describe what the rule should do:**
Flag assignments that are always overwritten (never read).

**What category of rule is this? (place an "X" next to just one item)**

[ ] Enforces code style
[X] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
var x = 1; // assigned value 1 is unused
x = 2;
console.log(x);
```

```js
let x = 1;
console.log(x);

x = 2; // assigned value 2 is unused
if (y) {
    x = 3;
}
else { // warning only because both branches overwrite the value
    x = 4;
}
console.log(x);
```

```js
// assigned value is unused, however f1() might have side effects
// so the message should not suggest removing the call, just removing the assignment.
var x = f1();
x = f2();
console.log(x);
```

**Why should this rule be included in ESLint (instead of a plugin)?**

This is general JS rule that is not related to any library.

It's also common in other language style checkers (Java, C#), as it might indicate a developer's error, or be misleading to other people reading the code. I've seen some "enterprise" JS checkers warning about that as well.

**Comment by Mouvedia** [comment]
Would it warn for an object as well? I am thinking about getters with side-effects.

**Comment by ashmind** [comment]
@Mouvedia This rule is more about unused assignment by itself, than a specific value.
It's like `no-unused-vars` looking at `var unused = o.sideeffect;` -- `no-unused-vars` will still complain even though there might be a side-effect.

So all of those are unused assignments (except the last):
```javascript
let x;

x = 1;
x = "string";
x = {};
x = y.p;
x = f();

x = "used";
console.log(x);
```

**Comment by Mouvedia** [comment]
Will you take into account conditions?
e.g.
```
let x = {}; // shouldn't throw

if (bar)
  x = foo();
else
  console.log('qux');

baz(x);
```

**Comment by ashmind** [comment]
@Mouvedia Yes, they should, some details below.

The following SHOULD NOT produce a warning:
```js
let x = 1;
if (c1) {
    x = 2;
}
else { // doesn't overwrite
    console.log('...');
}
```

```js
let x = 1;
if (true) { // always true, but analyzing the condition is out of scope
    x = 2;
}
```

```js
let x = 1;
switch (y) {
    case 'a': x = 2; break;
    case 'b': x = 3; break;
     // doesn't overwrite for other cases
}
```

```js
let x = 1;
for (const y of ys) { // no guarantee that ys has any items, so might not overwrite
    x = 2;
}
```

```js
let x = 1;
f(() => { x = 2; }); // we don't know when f() will call the callback, might not overwrite 1
```

```js
let x = 1;
f(() => x); // we don't know when f() will call the callback, might read 1
x = 2;
```

The following SHOULD produce a warning:
```js
let x = 1;
if (c1) {
    x = 2;
}
else if (c2) {
    x = 3;
}
else { // all branches overwrite
    x = 4;
}
```

```js
let x = 1;
switch (y) {
    case 'a': x = 2; break;
    case 'b': x = 3; break;
    default: x = 4; break; // all branches overwrite
}
```

```js
let x;
f(() => {
    x = 1; // we don't know when lambda is called, but within its body we can warn
    x = 2;
});
```

```js
let x = 1; // even though we don't know when functions are called, we can see that nothing uses 1
export function f1() { x = 2; /*...*/ }
export function f2() { x = 3; /*...*/ }
```

```js
let x = 1;
await f(); // even though some other code might run, nothing else can see `x`
x = 2;
```

```js
let x = 1;
yield y; // same as await
x = 2;
```

**Comment by Mouvedia** [comment]
Personally Id add an option to exclude functions and objects.
see https://github.com/eslint/eslint/issues/10885#issuecomment-424527823
